pkg/policy/validate_test.go

```diff
@@ -0,0 +1,203 @@
+  - keyless:
+      identities:
+      -  issuer: https://issuer.example.com
+         subject: foo@example.com
+      url: https://fulcio.sigstore.dev
+`,
+		wantErr: nil,
+	}, {
+		name: "good CIP and Secret",
+		doc: `
+apiVersion: policy.sigstore.dev/v1beta1
+kind: ClusterImagePolicy
+metadata:
+  name: blah
+spec:
+  images:
+  - glob: '*'
+  authorities:
+  - keyless:
+      identities:
+      -  issuer: https://issuer.example.com
+         subject: foo@example.com
+      url: https://fulcio.sigstore.dev
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: foo
+  namespace: cosign-system
+stringData:
+  foo: bar
+`,
+		wantErr: nil,
+	}, {
+		name: "bad secret namespace",
+		doc: `
+apiVersion: v1
+kind: Secret
+metadata:
+  name: foo
+  namespace: something-system
+stringData:
+  foo: bar
+`,
+		wantErr: errors.New(`invalid value: something-system: [0].metadata.namespace`),
+	}, {
+		name: "bad image policy",
+		doc: `
+apiVersion: policy.sigstore.dev/v1alpha1
+kind: ClusterImagePolicy
+metadata:
+  name: blah
+spec:
+  images:
+  - glob: '*'
+  authorities:
+  - key: {}
+`,
+		wantErr: apis.ErrMissingOneOf("data", "kms", "secretref").ViaField("key").ViaFieldIndex("authorities", 0).ViaField("spec"),
+	}, {
+		name:    "empty document",
+		doc:     ``,
+		wantErr: ErrEmptyDocument,
+	}, {
+		name: "object missing kind",
+		doc: `
+apiVersion: policy.sigstore.dev/v1beta1
+# Missing: kind: ClusterImagePolicy
+metadata:
+  name: blah
+spec: {}
+`,
+		wantErr: errors.New(`decoding object[0]: error unmarshaling JSON: while decoding JSON: Object 'Kind' is missing in '{"apiVersion":"policy.sigstore.dev/v1beta1","metadata":{"name":"blah"},"spec":{}}'`),
+	}, {
+		name: "unknown field",
+		doc: `
+apiVersion: policy.sigstore.dev/v1beta1
+kind: ClusterImagePolicy
+metadata:
+  name: blah
+spec:
+  asdf: dfsadf
+`,
+		wantErr: errors.New(`unable to unmarshal: json: unknown field "asdf"`),
+	}, {
+		name: "unknown type",
+		doc: `
+apiVersion: unknown.dev/v1
+kind: OtherPolicy
+metadata:
+  name: blah
+spec: {}
+`,
+		wantErr: errors.New(`unknown type: unknown.dev/v1, Kind=OtherPolicy`),
+	}, {
+		name: "warning - missing field",
+		doc: `
+apiVersion: policy.sigstore.dev/v1beta1
+kind: ClusterImagePolicy
+metadata:
+  name: blah
+spec:
+  images:
+  - glob: '*'
+  authorities:
+  - keyless:
+      url: https://fulcio.sigstore.dev
+`,
+		wantWarns: errors.New("missing field(s): spec.authorities[0].keyless.identities"),
+		wantErr:   nil,
+	},
+		{
+			name: "admit - missing authorities",
+			doc: `
+apiVersion: policy.sigstore.dev/v1beta1
+kind: ClusterImagePolicy
+metadata:
+  name: blah
+spec:
+  images:
+  - glob: '*'
+`,
+			wantErr:               nil,
+			allowEmptyAuthorities: true,
+		}, {
+			name: "deny - missing authorities",
+			doc: `
+apiVersion: policy.sigstore.dev/v1beta1
+kind: ClusterImagePolicy
+metadata:
+  name: blah
+spec:
+  images:
+  - glob: '*'
+`,
+			wantErr: errors.New("missing field(s): spec.authorities"),
```

Comment:
I'd prefer if we could reuse the YAMLs from the existing test/testdata for these tests rather than adding these here. But we can do this in a followup if you agree. They are used for validating the admission webhook, so would be nice to keep them in sync I think.
https://github.com/sigstore/policy-controller/tree/main/test/testdata/policy-controller
warn,invalid,valid, empty-authorities

Each of the warn, error has the expected failure baked into it, like so:
https://github.com/sigstore/policy-controller/blob/main/test/testdata/policy-controller/warn/v1beta1-missing-identity.yaml#L15

https://github.com/sigstore/policy-controller/blob/main/test/testdata/policy-controller/invalid/invalid-pubkey.yaml#L15

So thinking we would get more thorough testing and as we add more testdata files, we'll get those test cases for free.